### ISSUE TITLE
Performance of multiple (check-sat) calls in incremental mode

### ISSUE THREAD

**Issue by Tomaqa** [CVC4/CVC4#1718]
I was wondering why multiple (check-sat) calls in incremental mode do all take almost the same time even with the same assertion stack. I was expecting that the second call will ideally execute immediately, as no changes (no additional assertions) were performed. What reason is for this? And how can I achieve the effect that only changes in assertion stack are checked incrementally? Is it possible in the executable, only from C++ API, or not at all? Thanks.

**Comment by mpreiner** [comment]
Hi @Tomaqa,

CVC4 is not optimized for incremental usage yet. We are working on that. Can you please provide a benchmark on which CVC4 performs poorly in incremental mode?

**Comment by Tomaqa** [comment]
Here's one example from SMTLIB QF_UFLRA non-incremental benchmarks 'mathsat/RandomCoupled' with additional (check-sat) appended.
I'm not sure whether it's not a little overkill for you, I tried this one to check how it performs measurably. I think that any other smaller example from this set would do the job as well.
[pb_real_40_0800_10_18.txt](https://github.com/CVC4/CVC4/files/1855339/pb_real_40_0800_10_18.txt)

P.S. So is my assumption that from the API design point of view, it is alright to call (check-sat) multiple times with only small amout of changes in assertion stack? My task is to combine SMT and ODE solver by dynamically obtaining values from SMT, evaluating new ones in ODE and sending them back to SMT solver: so (check-sat) and (get-value) is necessary to call in each step.
Also I would like to stick at SMTLIB API so that any incremental SMTLIB-conform solver can be used.

(I have experienced the referenced issue also in other solvers.)

Thank you for your advice.
